File: 24-swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    first := head
    second := first.Next
    third := second.Next
    first.Next = swapPairs(third)
    second.Next = first
    return second
}

File: bwu/linked_list/24-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        curr = dummy
        curr.next = head
        
        while curr.next and curr.next.next:
            tmp = curr.next
            curr.next = tmp.next
            tmp.next = curr.next.next
            curr.next.next = tmp
            curr = tmp
        return dummy.next
